
import sys
import subprocess

input_bed = open(sys.argv[1])
input_arriba = open(sys.argv[2])
input_starfusion = open(sys.argv[3])
input_fusioncatcher = open(sys.argv[4])
input_bam = sys.argv[5]
output_fusions = open(sys.argv[6], "w")
output_coverage_file_name = sys.argv[7]


housekeeping_genes = ["GAPDH", "GUSB", "OAZ1", "POLR2A"]
artefact_genes = ["MAML2"]

output_fusions.write("Caller\tgene1\tgene2\texon1\texon2\tconfidence\tpredicted_effect\tbreakpoint1\tbreakpoint2\tcoverage1\tcoverage2\tsplit_reads1\tsplit_reads2\tSpanning_pairs\tsplit_reads\tBreakpoint1_covarage/SplitReads\tBreakpoint2_covarage/SplitReads\n")

#Only keep fusions with one gene that are in the design
design_genes_pool1 = {}
design_genes_pool2 = {}
design_genes = {}
for line in input_bed :
    lline = line.strip().split("\t")
    chrom = lline[0]
    start = int(lline[1])
    end = int(lline[2])
    exon = lline[3]
    gene = lline[3].split("_")[0]
    pool = lline[4]
    if gene in design_genes :
        design_genes[gene].append([chrom, start, end, exon])
    else :
        design_genes[gene] = [[chrom, start, end, exon]]
    if pool == 1 :
        design_genes_pool1[gene] = ""
    else :
        design_genes_pool2[gene] = ""

#Arriba fusions
header = True
for line in input_arriba :
    if header :
        header = False
        continue
    lline = line.strip().split("\t")
    gene1 = lline[0]
    gene2 = lline[1]
    #Only keep fusions with one gene that are in the design
    if (gene1 not in design_genes and gene2 not in design_genes) :
        continue
    confidence = lline[16]
    #Only keep fusions with high or medium confidence
    #if confidence == "low" :
    #    continue
    breakpoint1 = lline[4]
    breakpoint2 = lline[5]
    split_reads1 = lline[11]
    split_reads2 = lline[12]
    discordant_mates = lline[13]
    coverage1 = lline[14]
    coverage2 = lline[15]
    predicted_effect = lline[21]
    #Compare fusion coverage with coverage in breakpoints
    chrom1 = "chr" + breakpoint1.split(":")[0]
    pos1 = breakpoint1.split(":")[1]
    chrom2 = "chr" + breakpoint2.split(":")[0]
    pos2 = breakpoint2.split(":")[1]
    cov1 = 0
    cov2 = 0
    subprocess.call("samtools depth -d 50000 -a -r " + chrom1 + ":" + pos1 + "-" + pos1 + " " + input_bam + " > " + output_coverage_file_name, shell=True)
    output_coverage = open(output_coverage_file_name)
    for line in output_coverage :
        cov1 = int(line.strip().split("\t")[2])
    output_coverage.close()
    subprocess.call("samtools depth -d 50000 -a -r " + chrom2 + ":" + pos2 + "-" + pos2 + " " + input_bam + " > " + output_coverage_file_name, shell=True)
    output_coverage = open(output_coverage_file_name)
    for line in output_coverage :
        cov2 = int(line.strip().split("\t")[2])
    output_coverage.close()
    q1 = round((cov1 / (float(split_reads1) + float(split_reads2))),1)
    q2 = round((cov2 / (float(split_reads1) + float(split_reads2))),1)
    #Get exon name if it is in design
    exon1 = ""
    exon2 = ""
    if gene1 in design_genes :
        for region in design_genes[gene1] :
            if int(pos1) >= region[1] and int(pos1) >= region[2] :
                exon1 = region[3]
    if gene2 in design_genes :
        for region in design_genes[gene2] :
            if int(pos2) >= region[1] and int(pos2) >= region[2] :
                exon2 = region[3]
    output_fusions.write("Arriba\t" + gene1 + "\t" + gene2 + "\t" + exon1 + "\t" + exon2 + "\t" + confidence + "\t" + predicted_effect + "\t" + breakpoint1 + "\t" + breakpoint2 + "\t" + coverage1 + "\t" + coverage2 + "\t" + split_reads1 + "\t" + split_reads2 + "\t" + discordant_mates + "\t\t" + str(q1) + "\t" + str(q2) + "\n")


#Star-fusions
header = True
for line in input_starfusion :
    if header :
        header = False
        continue
    lline = line.strip().split("\t")
    gene1 = lline[0].split("--")[0]
    gene2 = lline[0].split("--")[1]
    #Only keep fusions with one gene that are in the design
    if (gene1 not in design_genes and gene2 not in design_genes) :
        continue
    Junction_read_count = lline[1]
    Spanning_Frag_count = lline[2]
    #Flag fusions with junction_read_count < 10 and Spanning_Frag_count < 2
    confidence = ""
    if int(Junction_read_count) < 15 :
        confidence = "Low support"
    #Remove Fusions with very weak read support
    if int(Junction_read_count) < 10 and int(Spanning_Frag_count) <= 2 :
        continue
    #Higher demand of read support for genes with frequent FP, house keeping genes, and pool2 genes without fusion to pool1 gene
    if (gene1 in artefact_genes or gene2 in artefact_genes or
        gene1 in housekeeping_genes or gene2 in housekeeping_genes or
        ((gene1 in design_genes_pool2 or gene2 in design_genes_pool2) and not (gene1 in design_genes_pool1 or gene2 in design_genes_pool1))) :
        if int(Junction_read_count) < 15 :
            continue
    breakpoint1 = lline[5]
    breakpoint2 = lline[7]
    FFPM = lline[9]
    DBs = lline[14]
    predicted_effect = lline[19]
    #Compare fusion coverage with coverage in breakpoints
    chrom1 = breakpoint1.split(":")[0]
    pos1 = breakpoint1.split(":")[1]
    chrom2 = breakpoint2.split(":")[0]
    pos2 = breakpoint2.split(":")[1]
    cov1 = 0
    cov2 = 0
    subprocess.call("samtools depth -d 50000 -a -r " + chrom1 + ":" + pos1 + "-" + pos1 + " " + input_bam + " > " + output_coverage_file_name, shell=True)
    output_coverage = open(output_coverage_file_name)
    for line in output_coverage :
        cov1 = int(line.strip().split("\t")[2])
    output_coverage.close()
    subprocess.call("samtools depth -d 50000 -a -r " + chrom2 + ":" + pos2 + "-" + pos2 + " " + input_bam + " > " + output_coverage_file_name, shell=True)
    output_coverage = open(output_coverage_file_name)
    for line in output_coverage :
        cov2 = int(line.strip().split("\t")[2])
    output_coverage.close()
    q1 = round((cov1 / (float(Junction_read_count))),1)
    q2 = round((cov2 / (float(Junction_read_count))),1)
    #Get exon name if it is in design
    exon1 = ""
    exon2 = ""
    if gene1 in design_genes :
        for region in design_genes[gene1] :
            if int(pos1) >= region[1] and int(pos1) >= region[2] :
                exon1 = region[3]
    if gene2 in design_genes :
        for region in design_genes[gene2] :
            if int(pos2) >= region[1] and int(pos2) >= region[2] :
                exon2 = region[3]
    output_fusions.write("StarFusion\t" + gene1 + "\t" + gene2 + "\t" + exon1 + "\t" + exon2 + "\t" + confidence + "\t" + predicted_effect + "\t" + breakpoint1 + "\t" + breakpoint2 + "\t\t\t\t\t" + Spanning_Frag_count + "\t" + Junction_read_count + "\t" + str(q1) + "\t" + str(q2) + "\n")



#FusionCatcher
header = True
for line in input_fusioncatcher :
    if header :
        header = False
        continue
    lline = line.strip().split("\t")
    gene1 = lline[0]
    gene2 = lline[1]
    #Only keep fusions with one gene that are in the design
    if (gene1 not in design_genes and gene2 not in design_genes) :
        continue
    fp_filters = lline[2].split(",")
    DBs = lline[2]
    common_mapping = lline[3]
    Spanning_pairs = lline[4]
    Spanning_reads_unique = lline[5]
    Fusion_finding_method = lline[7]
    breakpoint1 = lline[8]
    breakpoint2 = lline[9]
    predicted_effect = lline[15]
    #Flag fusions with Spanning_reads_unique < 5
    confidence = ""
    if int(Spanning_reads_unique) < 15:
        confidence = "Low support"
    #Filter fusions with very low support
    if int(Spanning_reads_unique) <= 5:
        continue
    #Higher demand of read support for genes with frequent FP, house keeping genes, and pool2 genes without fusion to pool1 gene
    if (gene1 in artefact_genes or gene2 in artefact_genes or
        gene1 in housekeeping_genes or gene2 in housekeeping_genes or
        ((gene1 in design_genes_pool2 or gene2 in design_genes_pool2) and not (gene1 in design_genes_pool1 or gene2 in design_genes_pool1))) :
        if int(Spanning_reads_unique) < 15 :
            continue
    #Flag fusions annotated that are fusions with very high probability
    fp_db = ["banned", "bodymap2", "cacg", "1000genomes", "conjoing", "cortex", "distance1000bp", "ensembl_fully_overlapping", "ensembl_same_strand_overlapping", "gtex", "hpa", "mt", "paralogs", "refseq_fully_overlapping", "refseq_same_strand_overlapping", "rrna", "similar_reads", "similar_symbols", "ucsc_fully_overlapping", "ucsc_same_strand_overlapping"]
    fp_found = ""
    for fp in fp_db :
        if fp in fp_filters :
            fp_found = "FP"
    #Compare fusion coverage with coverage in breakpoints
    if len(breakpoint1.split(":")) == 3 and len(breakpoint2.split(":")) == 3 :
        chrom1 = "chr" + breakpoint1.split(":")[0]
        pos1 = breakpoint1.split(":")[1]
        chrom2 = "chr" + breakpoint2.split(":")[0]
        pos2 = breakpoint2.split(":")[1]
        cov1 = 0
        cov2 = 0
        subprocess.call("samtools depth -d 50000 -a -r " + chrom1 + ":" + pos1 + "-" + pos1 + " " + input_bam + " > " + output_coverage_file_name, shell=True)
        output_coverage = open(output_coverage_file_name)
        for line in output_coverage :
            cov1 = int(line.strip().split("\t")[2])
        output_coverage.close()
        subprocess.call("samtools depth -d 50000 -a -r " + chrom2 + ":" + pos2 + "-" + pos2 + " " + input_bam + " > " + output_coverage_file_name, shell=True)
        output_coverage = open(output_coverage_file_name)
        for line in output_coverage :
            cov2 = int(line.strip().split("\t")[2])
        output_coverage.close()
        q1 = round((cov1 / (float(Spanning_reads_unique))),1)
        q2 = round((cov2 / (float(Spanning_reads_unique))),1)
    else :
        q1 = "NA"
        q2 = "NA"
    #Get exon name if it is in design
    exon1 = ""
    exon2 = ""
    if gene1 in design_genes :
        for region in design_genes[gene1] :
            if int(pos1) >= region[1] and int(pos1) >= region[2] :
                exon1 = region[3]
    if gene2 in design_genes :
        for region in design_genes[gene2] :
            if int(pos2) >= region[1] and int(pos2) >= region[2] :
                exon2 = region[3]
    #output_fusions.write("Caller\tgene1\tgene2\texon1\texon2\tconfidence\tpredicted_effect\tbreakpoint1\tbreakpoint2\tcoverage1\tcoverage2\tsplit_reads1\tsplit_reads2\tSpanning_pairs\tsplit_reads\tSpanning_unique_reads\tFusion_quotient1\tFusion_quotient2\n")
    #output_fusions.write("Arriba\t" + gene1 + "\t" + gene2 + "\t" + exon1 + "\t" + exon2 + "\t" + confidence + "\t" + predicted_effect + "\t" + breakpoint1 + "\t" + breakpoint2 + "\t" + coverage1 + "\t" + coverage2 + "\t" + split_reads1 + "\t" + split_reads2 + "\t" + discordant_mates + "\t\t" + str(q1) + "\t" + str(q2) + "\n")
    #output_fusions.write("StarFusion\t" + gene1 + "\t" + gene2 + "\t" + exon1 + "\t" + exon2 + "\t" + confidence + "\t\t" + breakpoint1 + "\t" + breakpoint2 + "\t\t\t\t\t" + Spanning_Frag_count + "\t" + Junction_read_count + "\t" + str(q1) + "\t" + str(q2) + "\n")
    output_fusions.write("FusionCatcher\t" + gene1 + "\t" + gene2 + "\t" + exon1 + "\t" + exon2 + "\t" + confidence + "\t" + predicted_effect + "\t" + breakpoint1 + "\t" + breakpoint2 + "\t\t\t\t\t" + Spanning_pairs + "\t" + Spanning_reads_unique + "\t" + str(q1) + "\t" + str(q2) + "\n")
